### PR TITLE
Added UncertaintyForest tutorial 4 out of 5: Running the learner

### LINKED ISSUE #188: Subtask of Issue 95: Write running tutorial notebook for UncertaintyForest
See #95

### PR DESCRIPTION
**Pull request by bstraus1**
Closes #188 

Added UncertaintyForest Tutorial 4/5: Running the learner

### CODE REVIEW

**Review by levinwil — CHANGES_REQUESTED** [review]
The accuracy is 0.5 - that is random chance. Something has to be wrong. Just because the data was generated from XOR randomly doesn’t mean the accuracy should be 0.5

Look at any other XOR tutorial or experiment and the results are not random chance.

**Comment by levinwil** [comment]
I’d try lowering d. I’d also try following the style of label_shuffle much more. 

This whole experiment series could literally just be this notebook alone, but heavily modified to be across multiple trials, with lower d, and with measures of calibration. 

Right now, this notebook series is just: showing how to install, which is already in the docs folder; the data setup, which is in the docs of whatever function you’re using (BTW there is XOR generation code in proglearn/sims)

This notebook just shows how to use the predict function, which is in the docs.

The tutorial should be “hey, we assume you know how to install this stuff and basic doc reading. This is a basic and isolated experiment to show how and why you’d use this package”

label_shuffle and the Hyppo tutorials are great examples of that

All of this is good without taking into context the larger repo and the stuff that exists in those files. But the first 3 notebooks are redundant information, and this notebook kind of is as well. 

I’d also give you the same comment I gave in the main chat: these notebooks shouldn’t have very much code at all. Most code should be defined in functions somewhere else in the package (e.g. proglearn/sims) or if they’re specific to this tutorial then define it in tutorials/functions, just as was done in label_shuffle

So I’d take a step back and don’t code for a bit. Think about what you’re trying to show. And then show it in the style of label_shuffle/hyppo.

**Comment by bstraus1** [comment]
First, I appreciate the feedback. Three things:

1. Even with lowering the d, my accuracy is 0.5. However, I feel that this is expected considering the training and testing data are completely random distinct datasets.
2. My plan is to create the tutorials with the code, then consolidate the code into a functions file as was done with label_shuffle
3. I included the step-by-step documentation because while the documentation for the repo is enough for the moderate/advanced user, it was not enough for me when I was starting out. I thought I'd make it easier on the next person who wants to run/edit/anything this algorithm without much background.

My plan going forward is as I mentioned in point 2. Is that okay with all?

**Comment by levinwil** [comment]
The test accuracy is so small (1) because the train size is 50 which is really small (2) because d = 100 which is huge. Please lower d to 2 and raise n_train to 10e5. You will likely see an increase in test accuracy.

This argument that “test accuracy will be random chance because they’re different datasets” violates every machine learning rule ever. Test set is always different than train set. If we only ever saw the training set then we’d just save it as a dictionary and get 100% accuracy. But we don’t. We look for patterns in train that will help us on test, if they’re drawn from the same distribution - which in this code, they are from the same distribution. And those patterns will hopefully give us better than random chance.  And they pretty much always do.

**Comment by levinwil** [comment]
With regards to the notebook series overall, the PR into staging should be the final single notebook like label_shuffle. Feel free to get there however you’d like. 

With regards to your notebooks on installation, we have an installation file now that wasn’t there before that shows how to install from the ground up. It’s in docs/install.rst

**Comment by bstraus1** [comment]
Okay, so I'll work on the tutorial series in my own fork. Then PR when it is all done.
